### PR TITLE
Pipeline tasks are only created in the database at runtime

### PR DESCRIPTION
**Pull request by kakj-go**
#### What type of this PR

/kind feature

#### What this PR does / why we need it:
In the automated test scenario, a large number of pipelines and tasks will be created. If they are created in the database before execution as before, it will cause a lot of pressure on the pipeline, so the creation process lags, and it is created during execution.

#### Which issue(s) this PR fixes:
erda-issue: [erda-issue](https://erda.cloud/erda/dop/projects/387/issues/all?id=58495&issueFilter__urlQuery=eyJzdGF0ZUJlbG9uZ3MiOlsiT1BFTiIsIldPUktJTkciLCJXT05URklYIiwiUkVPUEVOIiwiUkVTT0xWRUQiXSwiYXNzaWduZWVJRHMiOlsiMTAwMDU2MCJdfQ%3D%3D&issueTable__urlQuery=eyJwYWdlTm8iOjEsICJwYWdlU2l6ZSI6MTB9&issueViewGroup__urlQuery=eyJ2YWx1ZSI6ImthbmJhbiIsImNoaWxkcmVuVmFsdWUiOnsia2FuYmFuIjoiZGVhZGxpbmUifX0%3D&iterationID=506&type=BUG)

### CODE REVIEW

File: modules/pipeline/services/pipelinesvc/create.go [hunk 1]
```diff
@@ -189,199 +188,209 @@ func (s *PipelineSvc) makePipelineFromRequest(req *apistructs.PipelineCreateRequ
...
-// createPipelineGraph recursively create pipeline graph.
-// passedData stores data passed recursively.
-func (s *PipelineSvc) createPipelineGraph(p *spec.Pipeline, passedDataOpt ...passedDataWhenCreate) (err error) {
-	var passedData passedDataWhenCreate
-	if len(passedDataOpt) > 0 {
-		passedData = passedDataOpt[0]
-	}
-	passedData.initData(s.extMarketSvc)
+// traverse the stage of yml and save it to the database
+func (s *PipelineSvc) createPipelineGraphStage(p *spec.Pipeline, pipelineYml *pipelineyml.PipelineYml, txSession *xorm.Session) (stages []*spec.PipelineStage, err error) {
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/create.go:192]: use `dbclient.Session`, which wrapped `xorm.Session'

File: modules/pipeline/services/pipelinesvc/create.go [hunk 2]
```diff
@@ -189,199 +188,209 @@ func (s *PipelineSvc) makePipelineFromRequest(req *apistructs.PipelineCreateRequ
...
-		return apierrors.ErrCreatePipelineGraph.InternalError(err)
+	for si := range pipelineYml.Spec().Stages {
+		ps := &spec.PipelineStage{
+			PipelineID:  p.ID,
+			Name:        "",
+			Status:      apistructs.PipelineStatusAnalyzed,
+			CostTimeSec: -1,
+			Extra:       spec.PipelineStageExtra{StageOrder: si},
+		}
+		if err := s.dbClient.CreatePipelineStage(ps, dbclient.WithTxSession(txSession)); err != nil {
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/create.go:203]: passed with dbclient.SessionOptions

File: modules/pipeline/services/pipelinesvc/create.go [hunk 3]
```diff
@@ -189,199 +188,209 @@ func (s *PipelineSvc) makePipelineFromRequest(req *apistructs.PipelineCreateRequ
...
+			return nil, apierrors.ErrCreatePipelineGraph.InternalError(err)
+		}
+		dbStages = append(dbStages, ps)
 	}
-	defer func() {
+	return dbStages, nil
+}
+
+func (s *PipelineSvc) MergePipelineYmlTasks(pipelineYml *pipelineyml.PipelineYml, dbTasks []spec.PipelineTask, p *spec.Pipeline, dbStages []spec.PipelineStage, passedDataWhenCreate *task_extensions.PassedDataWhenCreate) (mergeTasks []spec.PipelineTask, err error) {
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/create.go:211]: Add comments.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/endpoints/permission.go [hunk 4]
```diff
@@ -61,27 +63,46 @@ func (e *Endpoints) checkBranchPermission(r *http.Request, appIDStr, branch stri
...
-	// 校验 task 是否属于当前 pipeline
-	tasks, err := e.dbClient.ListPipelineTasksByPipelineID(pipelineID)
+
+	yml, err := pipelineyml.New([]byte(p.PipelineYml))
+	if err != nil {
+		return apierrors.ErrCheckPermission.InvalidParameter(
+			fmt.Sprintf("yml parse error: %v", p.ID))
+	}
+
+	stages, err := e.dbClient.ListPipelineStageByPipelineID(p.ID)
```
[the remark this answers is not in the document]
Comment (kakj-go) [reply modules/pipeline/endpoints/permission.go:78]: Yes, it need to query all stages later

File: modules/pipeline/initialize.go [hunk 5]
```diff
@@ -131,7 +131,11 @@ func (p *provider) do() (*httpserver.Server, error) {
 	pipelineSvc.WithCmsService(p.CmsService)
 
 	pipelineFun := &reconciler.PipelineSvcFunc{
-		CronNotExecuteCompensate: pipelineSvc.CronNotExecuteCompensateById,
+		CronNotExecuteCompensate:                pipelineSvc.CronNotExecuteCompensateById,
```
[the remark this answers is not in the document]
Comment (kakj-go) [reply modules/pipeline/initialize.go:134]: Yes, I added todo and then consider how to refactor

File: modules/pipeline/pipengine/reconciler/define.go [hunk 6]
```diff
@@ -64,7 +68,11 @@ type Reconciler struct {
 
 // 该结构体为了解决假如 Reconciler 引入 pipelinesvc 导致循环依赖问题，所以将 svc 方法挂载进来
 type PipelineSvcFunc struct {
-	CronNotExecuteCompensate func(id uint64) error
+	CronNotExecuteCompensate                func(id uint64) error
```
[the remark this answers is not in the document]
Comment (kakj-go) [reply modules/pipeline/pipengine/reconciler/define.go:71]: Yes, I added todo and then consider how to refactor

File: modules/pipeline/pipengine/reconciler/reconcile.go [hunk 7]
```diff
@@ -35,22 +35,232 @@ import (
 	"github.com/erda-project/erda/modules/pipeline/spec"
 )
 
+func (r *Reconciler) MergeDBTasks(pipeline *spec.Pipeline) ([]*spec.PipelineTask, error) {
```
[the remark this answers is not in the document]
Comment (kakj-go) [reply modules/pipeline/pipengine/reconciler/reconcile.go:38]: Some methods require pipeline information

**Review by sfwn — COMMENTED** [review]
I will review it again tomorrow.

**Review by sfwn — APPROVED** [review]

**Pull request merged**